### PR TITLE
Interrupt all running tasks during shutdown

### PR DESCRIPTION
**Pull request by koppor**
(and don't allow new tasks to be executed)

This fixes https://github.com/jabref/jabref/issues/6109.

We invesitgated the shutdown procedure. We think that a thread creating a `.sav` file is still hanging around AFTER the throttler was shut down. And the `.sav` file was deleted afterwards. The `.sav` file was recreated after the deletion...

### CODE REVIEW

**Comment by stefan-kolb** [comment]
Not 100 % sure if this works, tho. We need to analyze our total executor architecture, I guess. Maybe we should alos document that in a markdown file as it is crucial for the application.

**Comment by calixtus** [comment]
Does this affect the WaitForSavedFinishedDialog?

**Comment by koppor** [comment]
Could also ref https://github.com/JabRef/jabref/pull/5967.

**Comment by koppor** [comment]
> Does this affect the WaitForSavedFinishedDialog?

No, because other sequence in the program flow. (Should I draw Sequence Diagrams? ^^)

**Comment by koppor** [comment]
> Should the same strategy also be used to shutdown the executors in JabRefExecutorService and DefaultTaskExecutor ?

Extended to `JabRefExecutorService`.

Extension to `DefaultTaskExecutor` seemed to be unnesseary as it works with `shutdownNow()` (and does not start with `shutdown()`.

File: src/main/java/org/jabref/JabRefExecutorService.java [hunk 1]
```diff
@@ -132,13 +132,47 @@ public void submit(TimerTask timerTask, long millisecondsDelay) {
         timer.schedule(timerTask, millisecondsDelay);
     }
 
+    /**
+     * Shuts everything down. Upon termination, this method returns.
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/JabRefExecutorService.java:136]: Upon termination, this method returns. 
??

File: src/main/java/org/jabref/JabRefExecutorService.java [hunk 2]
```diff
@@ -132,13 +132,47 @@ public void submit(TimerTask timerTask, long millisecondsDelay) {
...
+                LOGGER.debug("One minute passed, saving still not completed. Trying forced shutdown.");
+                executorService.shutdownNow();
+                if (executorService.awaitTermination(60, TimeUnit.SECONDS)) {
+                    LOGGER.debug("One minute passed again - forced shutdown worked.");
+                } else {
+                    LOGGER.error("DelayedTaskThrottler did not terminate");
+                }
+            }
+        } catch (InterruptedException ie) {
+                executorService.shutdownNow();
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/JabRefExecutorService.java:155]: indent

File: src/main/java/org/jabref/gui/util/DefaultTaskExecutor.java [hunk 3]
```diff
@@ -122,6 +122,9 @@ public static void runInJavaFXThread(Runnable runnable) {
         return scheduledExecutor.schedule(getJavaFXTask(task), delay, unit);
     }
 
+    /**
+     * Shuts everything down. Upon termination, this method returns.
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/util/DefaultTaskExecutor.java:126]: add to interface

File: src/main/java/org/jabref/JabRefExecutorService.java [hunk 4]
```diff
@@ -132,13 +132,47 @@ public void submit(TimerTask timerTask, long millisecondsDelay) {
...
+     */
     public void shutdownEverything() {
-        // those threads will be allowed to finish
-        this.executorService.shutdown();
-        // those threads will be interrupted in their current task
-        this.lowPriorityExecutorService.shutdownNow();
         // kill the remote thread
         stopRemoteThread();
+
+        try {
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/JabRefExecutorService.java:142]: extract try block to method and call it for executorService and lowPriorityExecutorService

File: src/main/java/org/jabref/JabRefExecutorService.java [hunk 5]
```diff
@@ -132,13 +132,47 @@ public void submit(TimerTask timerTask, long millisecondsDelay) {
...
+            // those threads will be allowed to finish
+            this.executorService.shutdown();
+            if (!executorService.awaitTermination(60, TimeUnit.SECONDS)) {
+                LOGGER.debug("One minute passed, saving still not completed. Trying forced shutdown.");
+                executorService.shutdownNow();
+                if (executorService.awaitTermination(60, TimeUnit.SECONDS)) {
+                    LOGGER.debug("One minute passed again - forced shutdown worked.");
+                } else {
+                    LOGGER.error("DelayedTaskThrottler did not terminate");
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/JabRefExecutorService.java:151]: thats not the delayed task throttler

File: src/main/java/org/jabref/logic/util/DelayTaskThrottler.java [hunk 6]
```diff
@@ -45,7 +46,25 @@ public void schedule(Runnable command) {
         }
     }
 
+    /**
+     * Shuts everything down. Upon termination, this method returns.
+     */
     public void shutdown() {
+        // this is non-blocking. See https://stackoverflow.com/a/57383461/873282.
         executor.shutdown();
+        try {
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/logic/util/DelayTaskThrottler.java:55]: call above helper method

**Review by tobiasdiez — CHANGES_REQUESTED** [review]

File: src/main/java/org/jabref/gui/util/DefaultTaskExecutor.java (hunk 3 above)
Comment (koppor) [reply src/main/java/org/jabref/gui/util/DefaultTaskExecutor.java:126]: Can't do - Added a comment to TaskExecutor

**Pull request merged**